src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs

```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+            
+            //Use legacy slow path if package does not implement IRenderPackageSupplement
+            if (packageSupplement == null)
+            {
+                LegacyPackageEnsureColorExistsPerVertex(package, previousPointVertexCount, previousLineVertexCount, previousMeshVertexCount);
+                return;
+            }
+            
+            if (package.PointVertexCount > previousPointVertexCount)
+            {
+                var count = package.PointVertexCount - packageSupplement.PointVertexColorCount;
+                if (count > 0)
+                {
+                    packageSupplement.AppendPointVertexColorRange(CreateColorByteArrayOfSize(count, DefR, DefG, DefB, DefA));
+                }
+                
+            }
 
-                    // The default color coming from the geometry library for
-                    // curves is 255,255,255,255 (White). Because we want a default
-                    // color of 0,0,0,255 (Black), we adjust the color components here.
-                    if (graphicItem is Curve || graphicItem is Surface || graphicItem is Solid || graphicItem is Point)
-                    {
-                        if (package.LineVertexCount > 0 && !package.LineStripVertexColors.Any())
-                        {
-                            package.ApplyLineVertexColors(CreateColorByteArrayOfSize(package.LineVertexCount, DefR, DefG, DefB, DefA));
-                        }
+            if (package.LineVertexCount > previousLineVertexCount)
+            {
+                var count = package.LineVertexCount - packageSupplement.LineVertexColorCount;
+                if (count > 0)
+                {
+                    packageSupplement.AppendLineVertexColorRange(CreateColorByteArrayOfSize(count, DefR, DefG, DefB, DefA));
+                }
+            }
 
-                        if (package.PointVertexCount > 0 && !package.PointVertexColors.Any())
-                        {
-                            package.ApplyPointVertexColors(CreateColorByteArrayOfSize(package.PointVertexCount, DefR, DefG, DefB, DefA));
-                        }
-                    }
+            if (package.MeshVertexCount > previousMeshVertexCount)
+            {
+                var count = package.MeshVertexCount - packageSupplement.MeshVertexColorCount;
+                if (count > 0)
+                {
+                    packageSupplement.AppendMeshVertexColorRange(CreateColorByteArrayOfSize(count, DefR, DefG, DefB, DefA));
                 }
-                catch (Exception e)
+            }
+        }
+
+        private static void LegacyPackageEnsureColorExistsPerVertex(IRenderPackage pkg, int previousPointVertexCount, int previousLineVertexCount, int previousMeshVertexCount)
+        {
+            if (pkg.PointVertexCount > previousPointVertexCount
+                && pkg.PointVertexColors.Any()
+                && pkg.PointVertexColors.Count() / 4 > pkg.PointVertexCount)
```

Comment:
Shouldn't this check be `pkg.PointVertexCount > pkg.PointVertexColors.Count() / 4`